src/chat/narrowsReducer.js

```diff
@@ -35,62 +35,76 @@ const messageFetchComplete = (state, action) => {
   const fetchedMessageIds = action.messages.map(message => message.id);
   const replaceExisting =
     action.anchor === FIRST_UNREAD_ANCHOR || action.anchor === LAST_MESSAGE_ANCHOR;
-  return {
-    ...state,
-    [key]: replaceExisting
+  return state.set(
+    key,
+    replaceExisting
       ? fetchedMessageIds
-      : union(state[key], fetchedMessageIds).sort((a, b) => a - b),
-  };
+      : union(state.get(key), fetchedMessageIds).sort((a, b) => a - b),
+  );
 };
 
 const eventNewMessage = (state, action) => {
-  let stateChange = false;
-  const newState: NarrowsState = {};
-  Object.keys(state).forEach(key => {
-    const { flags } = action.message;
-    if (!flags) {
-      throw new Error('EVENT_NEW_MESSAGE message missing flags');
-    }
-    const isInNarrow = isMessageInNarrow(action.message, flags, JSON.parse(key), action.ownEmail);
-    const isCaughtUp = action.caughtUp[key] && action.caughtUp[key].newer;
-    const messageDoesNotExist = state[key].find(id => action.message.id === id) === undefined;
-
-    if (isInNarrow && isCaughtUp && messageDoesNotExist) {
-      stateChange = true;
-      newState[key] = [...state[key], action.message.id];
-    } else {
-      newState[key] = state[key];
-    }
+  const { message } = action;
+  const { flags } = message;
+
+  if (!flags) {
+    throw new Error('EVENT_NEW_MESSAGE message missing flags');
+  }
+
+  return state.withMutations(stateMutable => {
+    const narrowsForMessage = getNarrowsForMessage(message, action.ownUser, flags);
+
+    narrowsForMessage.forEach(narrow => {
+      const key = JSON.stringify(narrow);
+      const value = stateMutable.get(key);
+
+      if (!value) {
+        // We haven't loaded this narrow. The time to add a new key
+        // isn't now; we do that in MESSAGE_FETCH_COMPLETE, when we
+        // might have a reasonably long, contiguous list of messages
+        // to show.
+        return; // i.e., continue
+      }
+
+      // Don't add a message to the end of the list unless we know
+      // it's the most recent message, i.e., unless we know we're
+      // currently looking at (caught up with) the newest messages in
+      // the narrow.
+      const isCaughtUp = action.caughtUp[key] && action.caughtUp[key].newer;
+      // Don't add a message that's already been added. It's probably
+      // very rare for a message to have already been added when we
+      // get an EVENT_NEW_MESSAGE, and perhaps impossible. (TODO:
+      // investigate?)
+      const messageDoesNotExist = value.find(id => action.message.id === id) === undefined;
+      if (isCaughtUp && messageDoesNotExist) {
```

Comment:
One subtle but potentially performance-significant wrinkle that's [in my fastest version](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.60Immutable.2EMap.60.20performance.20for.20.60state.2Enarrows.60/near/1067739):
> I tried another optimization: [graph]
> This is one I spotted because of staring at this code, but it's actually orthogonal to Immutable vs. objects-as-maps, and we should just do it in any case:
```diff
     for (const [key, value] of state.entries()) {
-      const isInNarrow = isMessageInNarrow(action.message, flags, JSON.parse(key), action.ownEmail);
-      const isCaughtUp = action.caughtUp[key] && action.caughtUp[key].newer;
-      const messageDoesNotExist = value.find(id => action.message.id === id) === undefined;
-
-      if (isInNarrow && isCaughtUp && messageDoesNotExist) {
+      if (
+        isMessageInNarrow(action.message, flags, JSON.parse(key), action.ownEmail)
+        && (action.caughtUp[key] && action.caughtUp[key].newer)
+        && value.find(id => action.message.id === id) === undefined
+      ) {
         stateMut.set(key, [...value, action.message.id]);
       }
     }
```
> I.e., no need to go looking through the whole narrow to see if the message is already there, if it's a narrow it doesn't even go in.

In general, the pattern this existing code is doing is a bad practice: computing several different complex or potentially-expensive conditions, and only at the end looking to see that the first one already gave us our answer and the rest were irrelevant. Instead, it's better to check each condition after computing it and act on it immediately. This probably isn't as big a deal for performance as it was in my measurements, because those were mostly reflecting the `isMessageInNarrow` condition, but best to not have to worry about it.

Probably the cleanest way to format that, in the presence of your helpful comments in the next commit, is to use the early-return pattern:
```js
      if (!(action.caughtUp[key] && action.caughtUp[key].newer)) {
        // Don't add a message to the end of the list unless we know
        // it's the most recent message, i.e., unless we know we're
        // currently looking at (caught up with) the newest messages in
        // the narrow.
        return;
      }
```
though a long chain of `&&` in one `if` statement (like in my demo version) is another time-honored alternative.

Reply:
Ah, right. I had read and considered that optimization, but I skipped it for readability with those new comments in mind (not having considered the quite readable early-return pattern, oops) since the number of iterations has decreased to a constant ~5, as you say:

> This probably isn't as big a deal for performance as it was in my measurements, because those were mostly reflecting the `isMessageInNarrow` condition

But you're right; EVENT_NEW_MESSAGE handling is a place where we should take any easy performance gains enthusiastically. 🙂